specs/core/1_shard-data-chains.md

```diff
@@ -333,10 +338,15 @@ def process_shard_block_header(beacon_state: BeaconState, shard_state: ShardStat
     # Verify the slot number
     assert block.slot == shard_state.slot
     # Verify the beacon chain root
-    parent_epoch = compute_epoch_of_shard_slot(shard_state.latest_block_header.slot)
-    assert block.beacon_block_root == get_block_root(beacon_state, parent_epoch)
+    epoch = compute_epoch_of_shard_slot(shard_state.slot)
+    assert epoch == compute_epoch_of_slot(beacon_state.slot)
+    if epoch * SLOTS_PER_EPOCH == beacon_state.slot:
```

Comment:
I'm tempted to force the beacon state to be at an epoch transition, i.e. `assert epoch * SLOTS_PER_EPOCH == beacon_state.slot`.

Reply:
This is already the case. What this code is doing is retrieving the block root when `slot % SLOTS_PER_EPOCH == 0`. At this type of slot, the block root is not yet in the state so you can't get it from `get_block_root`.  You have to get it from the `latest_block_header`.

actually, now that I think of it, you would need to be even more verbose and add the state root into the block header

```
header = deepcopy(beacon_state.latest_block_header)
header.state_root = hash_tree_root(beacon_state)
beacon_block_root = signing_root(header)
```

(can do it without the `deepcopy`. just for illustration)

Reply:
edit:
> I'm tempted to force the beacon state to be at an epoch transition, i.e. assert epoch * SLOTS_PER_EPOCH == beacon_state.slot.

it's the case because of the fact that `get_block_root` takes in an epoch and gives the start block of the epoch so the assertion that `block.beacon_block_root == beacon_block_root`

Reply:
While it is the case that the `get_block_root` is invariant within an epoch it's not necessarily the case for other pieces of beacon state. Just to be on the safe side I would force `shard_state_transition` to only accept the `BeaconState` if `beacon_state.slot % SLOTS_PER_EPOCH == 0`. It would also simplify the logic a bit in this specific instance (no need for `else` clause).

Reply:
ah, I see now. Will change it up

Reply:
Did this. You still have to do a conditional check on the `state_root == Bytes32` and fill in if necessary. (can't predict it ahead of time because there may or may not have been at block at the beacon slot in question)